src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,531 @@
+
+                return roleRequest;
+            }
+
+            private static HttpRequestMessage CreateTokenRequest(Uri baseUri)
+            {
+                var tokenRequest = new HttpRequestMessage
+                {
+                    RequestUri = new Uri(baseUri, "latest/api/token"),
+                    Method = HttpMethod.Put,
+                };
+                tokenRequest.Headers.Add("X-aws-ec2-metadata-token-ttl-seconds", "30");
+
+                return tokenRequest;
+            }
+
+            private static async Task<string> GetHttpContent(HttpRequestMessage request, string exceptionMessage)
+            {
+                HttpResponseMessage response;
+                try
+                {
+                    response = await __httpClientInstance.Value.SendAsync(request).ConfigureAwait(false);
+                    response.EnsureSuccessStatusCode();
+                }
+                catch (OperationCanceledException ex)
+                {
+                    throw new MongoClientException(exceptionMessage, ex);
+                }
+                catch (HttpRequestException ex)
+                {
+                    throw new MongoClientException(exceptionMessage, ex);
+                }
+
+                return await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+            }
+        }
+
+        private class MongoAWSMechanism : ISaslMechanism
+        {
+            private readonly UsernamePasswordCredential _credential;
+            private readonly IClock _dateTimeProvider;
+            private readonly IRandomByteGenerator _randomByteGenerator;
+            private readonly string _sessionToken;
+
+            public MongoAWSMechanism(
+                UsernamePasswordCredential credential,
+                string sessionToken,
+                IRandomByteGenerator randomByteGenerator,
+                IClock dateTimeProvider)
+            {
+                _credential = Ensure.IsNotNull(credential, nameof(credential));
+                _sessionToken = sessionToken;
+                _randomByteGenerator = Ensure.IsNotNull(randomByteGenerator, nameof(randomByteGenerator));
+                _dateTimeProvider = Ensure.IsNotNull(dateTimeProvider, nameof(dateTimeProvider));
+            }
+
+            public string Name
+            {
+                get { return MechanismName; }
+            }
+
+            public ISaslStep Initialize(IConnection connection, SaslConversation conversation, ConnectionDescription description)
+            {
+                Ensure.IsNotNull(connection, nameof(connection));
+                Ensure.IsNotNull(description, nameof(description));
+
+                var nonce = GenerateRandomBytes();
+
+                var document = new BsonDocument()
+                    .Add("r", new BsonBinaryData(nonce))
+                    .Add("p", new BsonInt32('n'));
+
+                var clientFirstMessageBytes = ToBytes(document);
```

Comment:
Why can't we just use:

    var clientMessageBytes = document.ToBson();

And similarly everywhere else ToBytes is called.